i18n-helpers/src/bin/markdown-xgettext.rs

```diff
@@ -0,0 +1,229 @@
+use anyhow::{bail, Context};
+use clap::{Arg, Command};
+use mdbook_i18n_helpers::{add_message, extract_messages};
+use polib::catalog::Catalog;
+use polib::metadata::CatalogMetadata;
+use polib::po_file;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+fn create_catalog(
+    content: &str,
+    source_path: &str,
+    language: &str,
+    title: Option<&str>,
+) -> anyhow::Result<Catalog> {
+    let mut metadata = CatalogMetadata::new();
+
+    if let Some(title) = title {
+        metadata.project_id_version = String::from(title);
+    }
+
+    metadata.language = String::from(language);
+    let now = chrono::Local::now();
+    metadata.pot_creation_date = now.to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
+    metadata.mime_version = String::from("1.0");
+    metadata.content_type = String::from("text/plain; charset=UTF-8");
+    metadata.content_transfer_encoding = String::from("8bit");
+    let mut catalog = Catalog::new(metadata);
+
+    for (lineno, msgid) in extract_messages(content) {
+        let source = format!("{}:{}", source_path, lineno);
+        add_message(&mut catalog, &msgid, &source);
+    }
+
+    Ok(catalog)
+}
+
+fn ingest_input(input_dir: Option<&str>) -> anyhow::Result<(&str, String)> {
+    if let Some(file_path) = input_dir {
+        let content = fs::read_to_string(file_path)
+            .context(format!("Failed to read from file {}", file_path))?;
+        Ok((file_path, content))
+    } else {
+        let mut buffer = String::new();
+        io::stdin()
+            .read_to_string(&mut buffer)
+            .context("Failed to read from stdin")?;
+        Ok(("-", buffer))
+    }
+}
+
+fn build_path(output_dir: Option<&str>, pot_title: &str) -> anyhow::Result<PathBuf> {
+    let mut output_path = PathBuf::new();
+
+    if let Some(out_dir) = output_dir {
+        let out_dir_path = Path::new(&out_dir);
+
+        if !out_dir_path.is_dir() {
+            bail!("The specified output path is not a directory.");
+        }
+        output_path.push(out_dir_path);
+    } else {
+        output_path.push(".");
+    }
+
+    output_path.push(pot_title);
+
+    Ok(output_path)
+}
+
+fn main() -> anyhow::Result<()> {
+    let matches = Command::new("markdown-xgettext")
+        .about("binary that extracts translatable text from markdown file for mdbook-i18n-helpers")
+        .arg(Arg::new("input_dir").short('f').long("file"))
+        .arg(Arg::new("language").short('l').long("lang").required(true))
+        .arg(Arg::new("pot_title").short('t').long("title"))
+        .arg(Arg::new("output_dir").short('o').long("out"))
+        .get_matches();
+
+    let (input, content) =
+        ingest_input(matches.get_one::<String>("input_dir").map(|d| d.as_str()))?;
```

Comment:
I don't know Clap well, but I think this can be done more directly: by using the `derive` feature, we can work with the command line arguments as a typed struct: https://docs.rs/clap/latest/clap/#example.

I think that will be nicer: the `String` and `as_str()` calls here can be removed.